cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	}
+	return pvsDetails.CRN, nil
+}
+
+// TODO(karthik-k-n): Decide on proper naming format for services.
+
+// GetServiceName returns name of given service type from spec or generate a name for it.
+func (s *PowerVSClusterScope) GetServiceName(resourceType ResourceType) *string {
+	switch resourceType {
+	case ServiceInstance:
+		if s.ServiceInstance() == nil || s.ServiceInstance().Name == nil {
+			return pointer.String(fmt.Sprintf("%s-serviceInstance", s.InfraCluster().GetName()))
+		}
+		return s.ServiceInstance().Name
+	case Network:
+		if s.Network().Name == nil {
+			return pointer.String(fmt.Sprintf("DHCPSERVER%s-dhcp_Private", s.InfraCluster().GetName()))
+		}
+		return s.Network().Name
+	case VPC:
+		if s.VPC() == nil || s.VPC().Name == nil {
+			return pointer.String(fmt.Sprintf("%s-vpc", s.InfraCluster().GetName()))
+		}
+		return s.VPC().Name
+	case TransitGateway:
+		if s.TransitGateway() == nil || s.TransitGateway().Name == nil {
+			return pointer.String(fmt.Sprintf("%s-transitgateway", s.InfraCluster().GetName()))
+		}
+		return s.TransitGateway().Name
+	case DHCPServer:
+		return pointer.String(fmt.Sprintf("%s-dhcp", s.InfraCluster().GetName()))
+	case COSInstance:
+		if s.COSInstance() == nil || s.COSInstance().Name == "" {
+			return pointer.String(fmt.Sprintf("%s-cosinstance", s.InfraCluster().GetName()))
+		}
+		return &s.COSInstance().Name
+	case Subnet:
+		return pointer.String(fmt.Sprintf("%s-vpcsubnet", s.InfraCluster().GetName()))
+	}
+	return nil
+}
+
+// DeleteLoadBalancer deletes loadBalancer.
+func (s *PowerVSClusterScope) DeleteLoadBalancer() error {
+	if !s.deleteResource(LoadBalancer) {
+		return nil
+	}
+
+	for _, lb := range s.IBMPowerVSCluster.Status.LoadBalancers {
+		if lb.ID != nil {
+			lb, _, err := s.IBMVPCClient.GetLoadBalancer(&vpcv1.GetLoadBalancerOptions{
+				ID: lb.ID,
+			})
+
+			if err != nil {
+				if strings.Contains(err.Error(), "cannot be found") {
+					return nil
+				}
+				return fmt.Errorf("error fetching the load balancer: %w", err)
+			}
+
+			if lb != nil && lb.ProvisioningStatus != nil && *lb.ProvisioningStatus != string(infrav1beta2.VPCLoadBalancerStateDeletePending) {
+				_, err = s.IBMVPCClient.DeleteLoadBalancer(&vpcv1.DeleteLoadBalancerOptions{
+					ID: lb.ID,
+				})
+				if err != nil {
+					s.Error(err, "error deleting the load balancer")
+					return err
+				}
+				s.Info("Load balancer successfully deleted")
+			}
+		}
+	}
+	return nil
+}
+
+// DeleteVPCSubnet deletes VPC subnet.
+func (s *PowerVSClusterScope) DeleteVPCSubnet() error {
+	if !s.deleteResource(Subnet) {
+		return nil
+	}
+
+	for _, subnet := range s.IBMPowerVSCluster.Status.VPCSubnet {
+		if subnet.ID == nil {
+			continue
+		}
+
+		net, _, err := s.IBMVPCClient.GetSubnet(&vpcv1.GetSubnetOptions{
+			ID: subnet.ID,
+		})
+
+		if err != nil {
+			if strings.Contains(err.Error(), "Subnet not found") {
+				return nil
+			}
+			return fmt.Errorf("error fetching the subnet: %w", err)
+		}
+
```

Comment:
I think you need to wait for load balancer to get deleted before proceeding to subnet deletion or else it will fail.
Please see whether we need a check here like how a check is there for dhcpserver before deleting powervs instance.

Reply:
The idea behind this was that on `reconcileDelete`, we are not waiting for one resource to get deleted before proceeding to the other resource but rather call delete API on all the resources and return an aggregate of errors at the end. One thing is it logs the error before a successful deletion in the subsequent reconciliations.

I think it might get redundant adding check for LB here again since we check it with `DeleteLoadBalancer` in the previous step.

> Please see whether we need a check here like how a check is there for dhcpserver before deleting powervs instance.

For deleting powers instance, the check was added because even though the delete API returns with success, the service instance is not getting deleted when checked on UI. We thought it would be better to wait and delete any resources within it before deleting the workspace.

Reply:
Got it, only concerns me is that the intermediate error status user would see in ibmpowervscluster obj during destroy.
One example could be in future if some other external program query for this obj on an interval, not sure it would be good to throw these intermediate errors.

Reply:
@dharaneeshvrd, will be handling it in next iteration.